Assets/Mirror/Runtime/QuaternionCompression.cs

```diff
@@ -0,0 +1,237 @@
+                {
+                    index = i;
+                    current = next;
+                }
+            }
+
+            return index;
+        }
+
+        static Vector3 GetSmallerDimensions(int largestIndex, Quaternion value)
+        {
+            float x = value.x;
+            float y = value.y;
+            float z = value.z;
+            float w = value.w;
+
+            float a, b, c;
+            switch (largestIndex)
+            {
+                case 0:
+                    a = y;
+                    b = z;
+                    c = w;
+                    break;
+                case 1:
+                    a = x;
+                    b = z;
+                    c = w;
+                    break;
+                case 2:
+                    a = x;
+                    b = y;
+                    c = w;
+                    break;
+                case 3:
+                    a = x;
+                    b = y;
+                    c = z;
+                    break;
+                default:
+                    return Vector3.zero;
+            }
+            // largest needs to be positive to be calculated by reader 
+            // if largest is negative flip sign of others
+            float largestSign = Mathf.Sign(value[largestIndex]);
+            return new Vector3(a, b, c) * largestSign;
+        }
+
+        static Quaternion FromSmallerDimensions(uint largestIndex, Vector3 smallest)
+        {
+            float a = smallest.x;
+            float b = smallest.y;
+            float c = smallest.z;
+
+            float x, y, z, w;
+            float largest = Mathf.Sqrt(1 - a * a - b * b - c * c);
+            switch (largestIndex)
+            {
+                case 0:
+                    x = largest;
+                    y = a;
+                    z = b;
+                    w = c;
+                    break;
+                case 1:
+                    x = a;
+                    y = largest;
+                    z = b;
+                    w = c;
+                    break;
+                case 2:
+                    x = a;
+                    y = b;
+                    z = largest;
+                    w = c;
+                    break;
+                case 3:
+                    x = a;
+                    y = b;
+                    z = c;
+                    w = largest;
+                    break;
+                default:
+                    return Quaternion.identity;
+            }
+            // 0.999999f is tolerance for Unity's Dot function
+            Debug.Assert(x * x + y * y + z * z + w * w < (2 - 0.999999f), "larger than 1");
+            return new Quaternion(x, y, z, w).normalized;
+        }
+
+        static uint ScaleToUInt(float value, int targetBitLength)
+        {
+            // same as Mathf.Pow(2, targetBitLength) - 1
+            int targetRange = (1 << targetBitLength) - 1;
+
+            if (value > MaxValue) { return (uint)targetRange; }
```

Comment:
this function relies on a static MaxValue.
and the target range is hard coded.

it would be cleaner if the function relies on no global state.
this way it can be reused whenever we need to scale a float to an uint.
for example:
```public static uint ScaleToUInt(float value, float minValue, float maxValue, uint minTarget, ushort maxTarget)```

Reply:
I've changed them and made them public, If we plan to add a lot of methods to this Compression class it might be better to instead do like c# Encoding class and have them nested, like `Compression.Quaternion.Compress(...)` `Compression.Float.Compress(...)` so that each type is Encapsulated and we have to worry less about using the wrong constant values.